sympy/matrices/expressions/matadd.py

```diff
@@ -52,12 +52,17 @@ def __new__(cls, *args, evaluate=False, check=False, _sympify=True):
             validate(*args)
 
         if evaluate:
-            if not any(isinstance(i, MatrixExpr) for i in args):
-                return Add(*args, evaluate=True)
-            obj = canonicalize(obj)
+            obj = cls._evaluate(obj)
 
         return obj
 
+    @classmethod
+    def _evaluate(cls, expr):
+        args = expr.args
+        if not any(isinstance(i, MatrixExpr) for i in args):
+            return Add(*args, evaluate=True)
+        return canonicalize(expr)
```

Comment:
Rather than testing for MatrixExpr it would be better to use `kind` here.

This looks like masking something like a bug though. Why would a MatAdd be constructed with non-matrix args?

This would make sense for a MatMul since a MatMul can always have some non-matrix args. There's never any reason to have any non-matrix args in a MatAdd though.

Reply:
I have just copied the previously existing code. I'm not sure about changing the behaviour of `MatAdd` in this PR.